### PR TITLE
add_subdomain_names without add_subdomain_ids

### PR DESCRIPTION
**Pull request by jmeier**
## Reason
Up until now, add_subdomain_names can only be used in combination with add_subdomain_ids. This forces the user to define subdomain ids by hand. To simplify the definition, it would be nice if Moose could simply assign a free ID for such additional subdomains automatically.

## Design
With this PR, Moose auto-assigns ids to subdomain names given in add_subdomain_names.

## Impact
This PR adds something new. No existing Moose model should be affected.

## Outlook
In case this PR is accepted, I'm happy to make a second, copy-pasta, one for add_sideset_names (see #27914).

### CODE REVIEW

File: test/tests/mesh/add_subdomain_ids/tests [hunk 1]
```diff
@@ -16,4 +16,12 @@
     design = MooseMesh.md
     requirement = 'The system shall allow specifying valid subdomain IDs and names that do not exist on the initial mesh.'
   []
+  [add_subdomain_names]
+    type = Exodiff
+    input = add_subdomain_names.i
+    exodiff = add_subdomain_names_out.e
+    issues = '#27968'
+    design = MooseMesh.md
+    requirement = 'The system shall allow specifying valid subdomain names that do not exist on the initial mesh.'
```
Comment (GiudGiud) [inline test/tests/mesh/add_subdomain_ids/tests:25]: add `exodiff_opts = '-pedantic'`
to make sure the assignements dont change

File: test/tests/mesh/add_subdomain_ids/add_subdomain_names.i [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    boundary = left
+    value = 0
+  []
+  [right]
+    type = DirichletBC
+    variable = u
+    boundary = right
+    value = 1
+  []
+[]
```
Comment (GiudGiud) [inline test/tests/mesh/add_subdomain_ids/add_subdomain_names.i:42]:
```suggestion
```

this is an antiquated way of creating mesh tests.
We now use `cli_args = '--mesh-only'` to create tests that only check the mesh

**Review by GiudGiud — CHANGES_REQUESTED** [review]
I think this is a great idea for this option

File: framework/src/mesh/MooseMesh.C [hunk 3]
```diff
@@ -432,9 +436,25 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
...
   }
   else if (isParamValid("add_subdomain_names"))
+  {
     // the user has defined add_subdomain_names, but not add_subdomain_ids
-    paramError("add_subdomain_names",
-               "In combination with add_subdomain_names, add_subdomain_ids must be defined.");
+    const auto & add_subdomain_names = getParam<std::vector<SubdomainName>>("add_subdomain_names");
+
+    // to define subdomain ids, we need the largest subdomain id defined yet.
+    int offset = 0;
```
Comment (GiudGiud) [inline framework/src/mesh/MooseMesh.C:444]:
```suggestion
    subdomain_id_type offset = 0; 
```

File: framework/src/mesh/MooseMesh.C [hunk 4]
```diff
@@ -432,9 +436,25 @@ MooseMesh::prepare(const MeshBase * const mesh_to_clone)
...
+
+    // to define subdomain ids, we need the largest subdomain id defined yet.
+    int offset = 0; 
+    if (!_mesh_subdomains.empty()) 
+        offset = *_mesh_subdomains.rbegin(); 
+
+    // add all subdomains (and auto-assign ids)
+    for (const SubdomainName & sub_name : add_subdomain_names)
+    {
+      int sub_id = offset++;
```
Comment (GiudGiud) [inline framework/src/mesh/MooseMesh.C:451]:
```suggestion
      const auto sub_id = offset++;
```

**Review by GiudGiud — COMMENTED** [review]
looks good to me.
Please clean the commit history:
- squashing update commits
- mention changes instead of mentioning reviewer comments which are not captured outside of gh website

**Comment by grunerjmeier** [comment]
> looks good to me. Please clean the commit history:
> 
> * squashing update commits
> * mention changes instead of mentioning reviewer comments which are not captured outside of gh website

Dear @GiudGiud,

I'll look into that. Please allow me to address one thing before (because it might add more commits to squash): Are you sure the test is working for our case? We create an empty subdomain and write an .e-file. Is the empty subdomain present in the .e-file? Paraview does not show it.

**Comment by GiudGiud** [comment]
It does not show on the mesh yet. It shows in paraview in the list of subdomains.
Does the text exodiff is that line is removed?

**Comment by grunerjmeier** [comment]
I did a small experiment with my test and I do not like the result:

When I remove "add_subdomain_names = 'MyAdditionalSubdomain'" from my test file "add_subdomain_names.i", the test is still passing (it should not). Regardless, if I have "exodiff_opts = '-pedantic'" or not.

In paraview "MyAdditionalSubdomain" is not showing up:
```
<?xml version="1.0"?>
<IOSS type="vtkDataAssembly" version="1.0" id="0">
  <element_blocks id="1">
    <block_0 id="2" label="block_0">
      <dataset id="0" />
    </block_0>
  </element_blocks>
  <node_sets id="3">
    <bottom id="4" label="bottom">
      <dataset id="1" />
    </bottom>
    <right id="5" label="right">
      <dataset id="2" />
    </right>
    <top id="6" label="top">
      <dataset id="3" />
    </top>
    <left id="7" label="left">
      <dataset id="4" />
    </left>
  </node_sets>
  <side_sets id="8">
    <bottom id="9" label="bottom">
      <dataset id="5" />
    </bottom>
    <right id="10" label="right">
      <dataset id="6" />
    </right>
    <top id="11" label="top">
      <dataset id="7" />
    </top>
    <left id="12" label="left">
      <dataset id="8" />
    </left>
  </side_sets>
</IOSS>
```

**Comment by grunerjmeier** [comment]
I managed to change the test. Now I'm struggeling with squashing the commits...

**Comment by GiudGiud** [comment]
See test failure
```
##########################################################################
ERROR: Your patch does not contain a valid ticket reference! (i.e. #1234)
Merge branch 'add_subdomain_names_only' of https://github.com/grunerjmeier/moose into test
allowing add_subdomain_names without add_subdomain_ids Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

ERROR: The following files do not contain a newline character before EOF:
	test/tests/mesh/add_subdomain_ids/add_subdomain_names.i

Run the "delete_trailing_whitespace.sh" script in your $MOOSE_DIR/scripts directory.
##########################################################################
```

You ll want to use VScode and set your user settings as shown here so you never run into this again
https://mooseframework.inl.gov/help/development/VSCode.html

**Comment by GiudGiud** [comment]
looks like we still have issues with the server test. I ll investigate

**Comment by GiudGiud** [comment]
The recover testing is showing a real failure
See https://civet.inl.gov/job/2293872/

It seems mesh->getSubdomainId with a new name is returning a new ID instead of the expected automatically assigned id. 

I tried a little fix. If it does not work you'll have to take care of it. I'm under water rn

**Review by GiudGiud — APPROVED** [review]
recover issue fixed. 
only contributed test fixes

**Comment by grunerjmeier** [comment]
Great! Thank you @GiudGiud!

**Pull request merged**